pkg/gc/gc_state_manager.go

```diff
@@ -0,0 +1,567 @@
+		if err1 != nil {
+			return err1
+		}
+
+		for _, barrier := range barriers {
+			if keyspaceID == constant.NullKeyspaceID && barrier.BarrierID == keypath.GCWorkerServiceSafePointID {
+				downgradeCompatibleMode = true
+				continue
+			}
+
+			if barrier.IsExpired(now) {
+				// Perform lazy delete to the expired GC barriers.
+				// WARNING: It might look like a reasonable optimization idea to perform the lazy-deletion in a lower
+				// frequency (instead of everytime checking it). However, it's UNSAFE considering the possibility of
+				// system clock drifts and PD leader changes, in which case an expired GC barrier may be back to
+				// not-expired state again. Once we regard a GC barrier as expired, it must be expired *strictly*,
+				// otherwise it may break the constraint that GC barriers must block the txn safe point from being
+				// advanced over them.
+				err1 = wb.DeleteGCBarrier(keyspaceID, barrier.BarrierID)
+				if err1 != nil {
+					return err1
+				}
+				// Do not block GC with expired barriers.
+				continue
+			}
+
+			if barrier.BarrierTS < minBlocker {
+				minBlocker = barrier.BarrierTS
+				blockingBarrier = barrier
+			}
+		}
+
+		// Compatible with old TiDB nodes that use TiDBMinStartTS to block GC.
+		ownerKey, minStartTS, err1 := m.gcMetaStorage.CompatibleLoadTiDBMinStartTS(keyspaceID)
+		if err1 != nil {
+			return err1
+		}
+
+		if minStartTS != 0 && len(ownerKey) != 0 && minStartTS < minBlocker {
+			// Note that txn safe point is defined inclusive: snapshots that exactly equals to the txn safe point are
+			// considered valid.
+			minBlocker = minStartTS
+			blockingBarrier = nil
+			blockingMinStartTSOwner = &ownerKey
+		}
+
+		// Txn safe point never decreases.
+		newTxnSafePoint = max(oldTxnSafePoint, minBlocker)
+
+		if downgradeCompatibleMode {
+			err1 = wb.SetGCBarrier(keyspaceID, endpoint.NewGCBarrier(keypath.GCWorkerServiceSafePointID, newTxnSafePoint, nil))
+			if err1 != nil {
+				return err1
+			}
+		}
+		return wb.SetTxnSafePoint(keyspaceID, newTxnSafePoint)
+	})
+	if err != nil {
+		return AdvanceTxnSafePointResult{}, err
+	}
+
+	blockerDesc := ""
+	if blockingBarrier != nil {
+		blockerDesc = blockingBarrier.String()
+	} else if blockingMinStartTSOwner != nil {
+		blockerDesc = fmt.Sprintf("TiDBMinStartTS { Key: %+q, MinStartTS: %d }", *blockingMinStartTSOwner, newTxnSafePoint)
+	}
+
+	if newTxnSafePoint != target {
+		if blockingBarrier == nil && blockingMinStartTSOwner == nil {
+			panic("unreachable")
+		}
+		if newTxnSafePoint == minBlocker {
+			log.Info("txn safe point advancement is being blocked",
+				zap.Uint64("oldTxnSafePoint", oldTxnSafePoint), zap.Uint64("target", target),
+				zap.Uint64("newTxnSafePoint", newTxnSafePoint), zap.String("blocker", blockerDesc),
+				zap.Bool("downgradeCompatibleMode", downgradeCompatibleMode))
+		} else {
+			log.Info("txn safe point advancement unable to be blocked by the minimum blocker",
```

Comment:
If I understand you correctly, when newTxnSafePoint != minBlocker, newTxnSafePoint = oldTxnSafePoint. So is this better?

```suggestion
			log.Info("txn safe point can't be advanced due to the minimum blocker is less than old txn safe point",
```

Reply:
Is this same as `txn safe point is remaining unchanged`?

Reply:
In this log I emphasized the fact that there is something that's expected to block the txn safe point at a smaller value, but the txn safe point has exceeded it already. In this case, the major risk is that some blocker isn't effective, instead of txn safe point not  being advanced.